File: mediator/java/GUIComponents.java
interface Mediator {

	void notify(Component sender, String msg);
}

abstract class Component {

	private Mediator mediator;

	protected Component(Mediator mediator) {
		this.mediator = mediator;
	}

	protected final void click() {
		this.mediator.notify(this, "Click");
	}
	
	protected final void keypress() {
		this.mediator.notify(this, "Keypress");
	}
}

final class Button extends Component {

	public Button(Mediator mediator) {
		super(mediator);
	}
}

final class Checkbox extends Component {

	public Checkbox(Mediator mediator) {
		super(mediator);
	}
}

final class Textbox extends Component {

	public Textbox(Mediator mediator) {
		super(mediator);
	}
}

final class ConcreteMediator implements Mediator {

	@Override
	public void notify(Component sender, String msg) {
		if (sender instanceof Button)
			System.out.println("[Button] "+msg);
		else if (sender instanceof Checkbox)
			System.out.println("[Checkbox] "+msg);
		else if (sender instanceof Textbox)
			System.out.println("[Textbox] "+msg);
	}
}

final class GUIComponents {

	public static void main(String[] args) {
		Mediator mediator = new ConcreteMediator();
		Component checkbox = new Checkbox(mediator);
		Component button = new Button(mediator);
		Component textbox = new Textbox(mediator);

		checkbox.click();
		checkbox.keypress();
		textbox.click();
		textbox.keypress();
		button.click();
		button.keypress();
	}
}

File: mediator/kotlin/GUI_components.kt
interface Mediator {

	fun notify(sender: Component, msg: String): Unit
}

abstract class Component(private val mediator: Mediator) {

	fun click(): Unit {
		mediator.notify(this, "Click")
	}

	fun keypress(): Unit {
		mediator.notify(this, "Keypress")
	}
}

class Button : Component {

	constructor(mediator: Mediator) : super(mediator)
}

class Textbox : Component {

	constructor(mediator: Mediator) : super(mediator)
}


class Checkbox : Component {

	constructor(mediator: Mediator) : super(mediator)
}

class ConcreteMediator : Mediator {

	override fun notify(sender: Component, msg: String): Unit {

		if (sender is Button) println("[Button] $msg")
		else if (sender is Textbox) println("[Textbox] $msg")
		else if (sender is Checkbox) println("[Checkbox] $msg")
	}
}


fun main(): Unit {
	val mediator: Mediator = ConcreteMediator()
	val button: Component = Button(mediator)
	val checkbox: Component = Checkbox(mediator)
	val textbox: Component = Textbox(mediator)

	button.click();
	button.keypress()
	checkbox.click();
	checkbox.keypress()
	textbox.click();
	textbox.keypress()
}